compiler/rustc_data_structures/src/sync.rs

```diff
@@ -407,54 +611,79 @@ impl<K: Eq + Hash, V: Eq, S: BuildHasher> HashMapExt<K, V> for HashMap<K, V, S>
 }
 
 #[derive(Debug)]
-pub struct Lock<T>(InnerLock<T>);
+pub struct Lock<T> {
+    single_thread: bool,
+    inner: RefCell<T>,
+    mt_inner: Option<Mutex<T>>,
+}
```

Comment:
I have strong reservations about this whole approach of having one synchronization mechanism for single-threaded and another for multi-threaded.

Putting that aside for a moment: is it possible to make this type an enum?
```
pub enum Lock<T> {
    Single(RefCell<T>),
    Multi(Mutex<T>),
}
```
The type would be smaller, simpler, and less error-prone.

Reply:
I tried this method at the beginning, but the effect looks very unsatisfactory:
https://github.com/rust-lang/rust/pull/101566#issuecomment-1241293886
and it is not even as good as using `std::sync::Mutex` directly:
https://github.com/rust-lang/rust/pull/101566#issuecomment-1276331871

Reply:
This seems to be caused by the optimization problem of `LockGuard`. If we can research a way to get better optimization(@Zoxc has a exprimental branch for it), I think it is feasible.

Reply:
What is the problem with `LockGuard`? #101566 is very long so it is hard for me to see.

Reply:
It's the comment https://github.com/rust-lang/rust/pull/101566#issuecomment-1407352906

Reply:
From my local test, it is mainly the `drop` function of `LockGuard` that affects the optimization of llvm